mysqltest/docker.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return 0, nopCloser, fmt.Errorf("dockertest.NewPool(): %w", err)
+	}
+	container, err := runContainer(conf, pool)
+	if err != nil {
+		return 0, nopCloser, err
+	}
+	closer := newContainerCloser(pool, container)
+	if err := container.Expire(uint(conf.killAfterSec)); err != nil {
+		return 0, closer, fmt.Errorf("resource.Expire(): %w", err)
+	}
+	outPort, err := waitUntilUp(conf, pool, container)
+	if err != nil {
+		return 0, closer, err
+	}
+	return outPort, closer, nil
+}
+
+// Starts the mysql container and returns a Resource that points to it.
+func runContainer(conf *config, pool *dockertest.Pool) (*dockertest.Resource, error) {
+	// pulls an image, creates a container based on it and runs it
+	container, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "mysql",
+		Tag:        conf.mySQLVersion,
+		Env:        []string{"MYSQL_ROOT_PASSWORD=" + password},
+	}, func(config *docker.HostConfig) {
+		config.AutoRemove = true // remove storage after container exits
+		config.RestartPolicy = docker.RestartPolicy{Name: "no"}
+	})
+	if err != nil {
+		var extraMsg string
+		switch {
+		case strings.Contains(err.Error(), "no such file"):
+			extraMsg = `. Please install docker: 
+		Instructions for gLinux: https://g3doc.corp.google.com/cloud/containers/g3doc/glinux-docker/install.md
+		Instructions for Debian: https://docs.docker.com/engine/install/debian/
+		Instructions for Mac: https://docs.docker.com/desktop/mac/install/`
+		case strings.Contains(err.Error(), "permission denied"):
+			extraMsg = `. To fix this, enable sudo-less docker container creation:
+					1. Run "sudo adduser $USER docker" to add your user to the docker group
+					2. Reboot the machine to make the group membership effective`
+		}
+		return nil, fmt.Errorf("pool.Run() failed starting mysql container: %w%s", err, extraMsg)
+	}
+	return container, nil
+}
+
+// waitUntilUp waits for mysql to be reachable.
+func waitUntilUp(conf *config, pool *dockertest.Pool, container *dockertest.Resource) (int, error) {
+	var outPort int
+	// To get the TCP port number for the mysql server, we have to wait for the docker container to
+	// actually start, then get the mapped port number.
+	pool.MaxWait = time.Minute
+	if err := pool.Retry(func() error {
+		p, err := fetchPort(container)
+		if err != nil {
+			return err
+		}
+
+		if err := tryLogin(conf, p); err != nil {
+			conf.progressLogger.Printf("MySQL isn't ready yet: %v", err)
+			return fmt.Errorf("MySQL isn't ready yet: %w", err)
+		}
+
+		outPort = p
+		return nil
+	}); err != nil {
+		return 0, fmt.Errorf("failed to connect to database within timeout. The final attempt returned: %w", err)
+	}
+	return outPort, nil
+}
+
+func fetchPort(r *dockertest.Resource) (int, error) {
+	portStr := r.GetPort("3306/tcp")
+	if portStr == "" {
+		return 0, fmt.Errorf("resource.GetPort() returned empty string, container isn't ready yet")
+	}
+	port, err := strconv.Atoi(portStr)
```

Comment:
Why do we need the port to be an integer?

Reply:
Changed to string throughout. I had forgotten that ports as strings were even a thing 😄